File: rust/periph/src/chips/io_expander/tpic6b595.rs
//! TPIC6B595 8-bit power SIPO shift register driver.
//!
//! Drives up to `num_devices` cascaded TPIC6B595s through a SiPo (serial-
//! in/parallel-out) connection. Each device exposes 8 open-drain outputs
//! (DRAIN0–DRAIN7); every write shifts the entire cascade MSB-first and pulses
//! RCK to latch all outputs atomically. Outputs only sink current — they never
//! source it; an external pull-up or load supply is required for the "off"/
//! high state.
//!
//! The driver owns a `num_devices`-byte shadow register so single-pin updates
//! work without re-reading the bus. Every write (pin, port, fill, write_all)
//! rebuilds and retransmits the entire reversed cascade — see
//! `specs/io_expander/tpic6b595.md` for the wire-order reversal that cascading
//! requires.
//!
//! Initialises every output to OFF at construction (shadow zero, latched
//! once). If the SiPo connection has SRCLR wired, the constructor pulses it
//! to clear the shift register before the all-zero latch.
//!
//! ## Safety
//!
//! The SiPo connection is owned through a shared reference. Multiple
//! [`ExPin`] objects may coexist, but simultaneous access from different ISR
//! contexts is not safe. Use only from a single execution context.

use core::cell::Cell;
use embedded_hal::digital::{ErrorKind, ErrorType, OutputPin, StatefulOutputPin};
use embedded_hal::spi::SpiBus;

use crate::connection::sipo::{SiPoConnection, SiPoError};

/// Wraps a SiPo error so it satisfies `embedded_hal::digital::Error`.
#[derive(Debug)]
pub struct PinError<E>(pub E);

impl<E: embedded_hal::spi::Error> embedded_hal::digital::Error for PinError<E> {
    fn kind(&self) -> ErrorKind { ErrorKind::Other }
}

/// Maximum cascade depth supported by the driver's shadow buffer.
pub const MAX_DEVICES: usize = 8;

/// TPIC6B595 minimal driver — exposes 8 × `num_devices` output pins as
/// [`ExPin`] GPIO proxies.
pub struct Tpic6b595Minimal<SPI, RCK, SRCLR, G> {
    sipo:       SiPoConnection<SPI, RCK, SRCLR, G>,
    num_devices: u8,
    shadow:     [Cell<u8>; MAX_DEVICES],
}

impl<SPI, RCK, SRCLR, G> Tpic6b595Minimal<SPI, RCK, SRCLR, G>
where
    SPI:  SpiBus,
    RCK:  embedded_hal::digital::OutputPin,
    SRCLR: embedded_hal::digital::OutputPin,
    G:    embedded_hal::digital::OutputPin,
{
    /// Create a new `Tpic6b595Minimal` and clear every output.
    ///
    /// # Arguments
    /// * `spi`        — Hardware or bit-banged SPI bus for SER IN / SRCK.
    /// * `rck`        — Pin connected to RCK (register clock).
    /// * `srclr`      — Pin connected to SRCLR, or `None` to disable it.
    /// * `g`          — Pin connected to G, or `None` to disable it.
    /// * `num_devices` — Number of cascaded TPIC6B595s on the wire (1–8).
    ///
    /// # Errors
    ///
    /// Returns whatever [`SiPoConnection::new`] returns (e.g. a GPIO error
    /// driving RCK/SRCLR/G to their idle states). Construction never
    /// returns an error from `SRCLR not configured` — a missing SRCLR line
    /// is silently ignored, since the chip is fully usable without it
    /// (the construction's `clear()` call is best-effort, and the all-zero
    /// write that follows guarantees every output starts OFF).
    pub fn new(
        spi: SPI,
        rck: RCK,
        srclr: Option<SRCLR>,
        g: Option<G>,
        num_devices: u8,
    ) -> Result<Self, SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        let n = if (num_devices as usize) > MAX_DEVICES { MAX_DEVICES as u8 } else { num_devices };
        let mut shadow = [Cell::new(0u8); MAX_DEVICES];
        for cell in shadow.iter_mut() { cell.set(0); }

        let sipo = SiPoConnection::new(spi, rck, srclr, g)?;
        // Best-effort clear of the shift register; missing SRCLR is fine.
        let _ = sipo.clear();

        let chip = Self { sipo, num_devices: n, shadow };
        chip.flush().ok();
        Ok(chip)
    }

    fn flush(&self) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        let mut wire = [0u8; MAX_DEVICES];
        for i in 0..(self.num_devices as usize) {
            wire[i] = self.shadow[self.num_devices as usize - 1 - i].get();
        }
        self.sipo.write(&wire[..self.num_devices as usize])
    }

    fn set_pin(&self, n: u8, high: bool) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        let port_idx = (n / 8) as usize;
        let bit = n % 8;
        let mut s = self.shadow[port_idx].get();
        if high { s |=   1 << bit; }
        else    { s &= !(1 << bit); }
        self.shadow[port_idx].set(s);
        self.flush()
    }

    /// Write all 8 outputs of cascaded device `port` from `mask`.
    ///
    /// `mask` bit 0 = DRAIN0, bit 7 = DRAIN7. 1 = ON (DMOS conducting,
    /// sinks current); 0 = OFF (high-impedance).
    pub fn write_port(&self, port: usize, mask: u8) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        self.shadow[port].set(mask);
        self.flush()
    }

    /// Set every pin on every cascaded device to `value`.
    pub fn fill(&self, value: bool) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        let b = if value { 0xFF } else { 0x00 };
        for cell in self.shadow.iter() { cell.set(b); }
        self.flush()
    }

    /// Turn every output off (equivalent to `fill(false)`).
    pub fn off(&self) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        self.fill(false)
    }

    /// Return an [`ExPin`] proxy for global pin `n` (0..`num_devices` * 8 - 1).
    pub fn pin(&self, n: u8) -> ExPin<'_, SPI, RCK, SRCLR, G> {
        ExPin { chip: self, n }
    }

    /// Return the shadow byte for the given port (no bus read).
    pub(crate) fn shadow_byte(&self, port: usize) -> u8 {
        self.shadow[port].get()
    }
}

/// GPIO proxy for a single TPIC6B595 pin — output-only.
pub struct ExPin<'a, SPI, RCK, SRCLR, G> {
    chip: &'a Tpic6b595Minimal<SPI, RCK, SRCLR, G>,
    n:    u8,
}

impl<SPI, RCK, SRCLR, G> ErrorType for ExPin<'_, SPI, RCK, SRCLR, G>
where
    SPI:  SpiBus,
    RCK:  embedded_hal::digital::OutputPin,
    SRCLR: embedded_hal::digital::OutputPin,
    G:    embedded_hal::digital::OutputPin,
{
    type Error = SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>;
}

impl<SPI, RCK, SRCLR, G> OutputPin for ExPin<'_, SPI, RCK, SRCLR, G>
where
    SPI:  SpiBus,
    RCK:  embedded_hal::digital::OutputPin,
    SRCLR: embedded_hal::digital::OutputPin,
    G:    embedded_hal::digital::OutputPin,
{
    fn set_high(&mut self) -> Result<(), Self::Error> {
        self.chip.set_pin(self.n, true)
    }

    fn set_low(&mut self) -> Result<(), Self::Error> {
        self.chip.set_pin(self.n, false)
    }
}

impl<SPI, RCK, SRCLR, G> StatefulOutputPin for ExPin<'_, SPI, RCK, SRCLR, G>
where
    SPI:  SpiBus,
    RCK:  embedded_hal::digital::OutputPin,
    SRCLR: embedded_hal::digital::OutputPin,
    G:    embedded_hal::digital::OutputPin,
{
    fn is_set_high(&mut self) -> Result<bool, Self::Error> {
        let port = (self.n / 8) as usize;
        let bit  = self.n % 8;
        Ok((self.chip.shadow_byte(port) >> bit) & 1 == 1)
    }

    fn is_set_low(&mut self) -> Result<bool, Self::Error> {
        Ok(!self.is_set_high()?)
    }
}

/// TPIC6B595 full driver — extends [`Tpic6b595Minimal`] with hardware features.
pub struct Tpic6b595Full<SPI, RCK, SRCLR, G> {
    inner: Tpic6b595Minimal<SPI, RCK, SRCLR, G>,
}

impl<SPI, RCK, SRCLR, G> Tpic6b595Full<SPI, RCK, SRCLR, G>
where
    SPI:  SpiBus,
    RCK:  embedded_hal::digital::OutputPin,
    SRCLR: embedded_hal::digital::OutputPin,
    G:    embedded_hal::digital::OutputPin,
{
    /// Create a new `Tpic6b595Full` and clear every output.
    pub fn new(
        spi: SPI,
        rck: RCK,
        srclr: Option<SRCLR>,
        g: Option<G>,
        num_devices: u8,
    ) -> Result<Self, SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        Ok(Self {
            inner: Tpic6b595Minimal::new(spi, rck, srclr, g, num_devices)?,
        })
    }

    /// Return an [`ExPin`] proxy for global pin `n`.
    pub fn pin(&self, n: u8) -> ExPin<'_, SPI, RCK, SRCLR, G> {
        self.inner.pin(n)
    }

    /// Write all 8 outputs of cascaded device `port` from `mask`.
    pub fn write_port(&self, port: usize, mask: u8) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        self.inner.write_port(port, mask)
    }

    /// Set every pin on every cascaded device to `value`.
    pub fn fill(&self, value: bool) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        self.inner.fill(value)
    }

    /// Turn every output off.
    pub fn off(&self) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        self.inner.off()
    }

    /// Pulse SRCLR to clear the shift register only.
    ///
    /// The storage register (and therefore the DRAIN outputs) keeps its
    /// last-latched value until the next RCK pulse.
    pub fn clear(&mut self) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        self.inner.sipo.clear()
    }

    /// Drive G LOW (`enabled = true`) or HIGH (`enabled = false`).
    pub fn set_output_enable(&mut self, enabled: bool) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        self.inner.sipo.set_output_enable(enabled)
    }

    /// Write every cascaded device's byte in one call.
    ///
    /// `values` is length-truncated or zero-extended to `num_devices` as needed.
    pub fn write_all(&mut self, values: &[u8]) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        for i in 0..(self.inner.num_devices as usize) {
            let v = if i < values.len() { values[i] } else { 0 };
            self.inner.shadow[i].set(v);
        }
        self.inner.flush()
    }
}

File: go/periph/chips/io_expander/tpic6b595.go
package ioexpander

import "fmt"

// TPIC6B595Minimal is the 8-bit power SIPO shift register driver —
// minimal interface.
//
// Drives up to numDevices cascaded TPIC6B595s through a SiPo
// (serial-in/parallel-out) connection. Each device exposes 8 open-drain
// outputs (DRAIN0–DRAIN7); every write shifts the entire cascade
// MSB-first and pulses RCK to latch all outputs atomically. Outputs
// only sink current — they never source it; an external pull-up or
// load supply is required for the "off"/high state.
//
// The driver owns a numDevices-byte shadow register so single-pin
// updates work without re-reading the bus. Every write (pin, port, fill)
// rebuilds and retransmits the entire reversed cascade — see
// specs/io_expander/tpic6b595.md for the wire-order reversal that
// cascading requires.
//
// Initialises every output to OFF at construction (shadow zero, latched
// once). If the SiPo connection has SRCLR wired, the constructor
// pulses it to clear the shift register before the all-zero latch.
type TPIC6B595Minimal struct {
	sipo        SiPo
	numDevices  uint8
	shadow      [8]uint8 // up to 8 cascaded devices
}

// MaxTPIC6B595Devices is the maximum cascade depth supported by the
// driver's shadow buffer.
const MaxTPIC6B595Devices = 8

// NewTPIC6B595Minimal creates a new TPIC6B595Minimal and clears every
// output.
//
// sipo must be a configured SiPo connection (a *connection.SIPOConnection
// from periph/connection). numDevices is the number of cascaded
// TPIC6B595s on the wire (1–8).
func NewTPIC6B595Minimal(sipo SiPo, numDevices uint8) (*TPIC6B595Minimal, error) {
	if numDevices == 0 || numDevices > MaxTPIC6B595Devices {
		return nil, fmt.Errorf("tpic6b595: numDevices must be 1..%d, got %d", MaxTPIC6B595Devices, numDevices)
	}
	d := &TPIC6B595Minimal{sipo: sipo, numDevices: numDevices}
	// Best-effort clear of the shift register; missing SRCLR is fine.
	_ = d.sipo.Clear()
	if err := d.flush(); err != nil {
		return nil, err
	}
	return d, nil
}

// WritePort writes all 8 outputs of cascaded device port from mask.
// mask bit 0 = DRAIN0, bit 7 = DRAIN7. 1 = ON (DMOS conducting, sinks
// current); 0 = OFF (high-impedance).
func (d *TPIC6B595Minimal) WritePort(port uint8, mask uint8) error {
	d.shadow[port] = mask
	return d.flush()
}

// Fill sets every pin on every cascaded device to value.
func (d *TPIC6B595Minimal) Fill(value bool) error {
	b := uint8(0x00)
	if value {
		b = 0xFF
	}
	for i := uint8(0); i < d.numDevices; i++ {
		d.shadow[i] = b
	}
	return d.flush()
}

// Off turns every output off (equivalent to Fill(false)).
func (d *TPIC6B595Minimal) Off() error {
	return d.Fill(false)
}

// Pin returns a TPIC6B595Pin proxy for global pin n (0..numDevices*8 - 1).
func (d *TPIC6B595Minimal) Pin(n uint8) TPIC6B595Pin {
	return TPIC6B595Pin{chip: d, n: n}
}

// Shadow returns the shadow byte for the given cascaded device port.
// Test-only — production callers should use Pin / Get / Set instead.
func (d *TPIC6B595Minimal) Shadow(port uint8) uint8 {
	return d.shadow[port]
}

// flush builds the reversed wire-order buffer from the shadow register
// and writes it via the SiPo connection.
func (d *TPIC6B595Minimal) flush() error {
	wire := make([]byte, d.numDevices)
	for i := uint8(0); i < d.numDevices; i++ {
		wire[i] = d.shadow[d.numDevices-1-i]
	}
	return d.sipo.Write(wire)
}

// TPIC6B595Pin is a GPIO proxy for a single TPIC6B595 pin — output-only.
type TPIC6B595Pin struct {
	chip *TPIC6B595Minimal
	n    uint8
}

// Get returns the shadow bit for this pin (NOT a bus read — SiPo is
// write-only). Returns 0 or 1.
func (p TPIC6B595Pin) Get() (bool, error) {
	port := p.n >> 3
	bit := p.n & 7
	return (p.chip.shadow[port]>>bit)&1 == 1, nil
}

// Set drives the DMOS output ON (high=true) or OFF (high=false).
func (p TPIC6B595Pin) Set(high bool) error {
	port := p.n >> 3
	bit := p.n & 7
	if high {
		p.chip.shadow[port] |= 1 << bit
	} else {
		p.chip.shadow[port] &^= 1 << bit
	}
	return p.chip.flush()
}

// Toggle inverts the shadow bit for this pin.
func (p TPIC6B595Pin) Toggle() error {
	v, err := p.Get()
	if err != nil {
		return err
	}
	return p.Set(!v)
}

// TPIC6B595Full extends TPIC6B595Minimal with hardware features.
type TPIC6B595Full struct {
	*TPIC6B595Minimal
}

// NewTPIC6B595Full creates a new TPIC6B595Full and clears every output.
func NewTPIC6B595Full(sipo SiPo, numDevices uint8) (*TPIC6B595Full, error) {
	min, err := NewTPIC6B595Minimal(sipo, numDevices)
	if err != nil {
		return nil, err
	}
	return &TPIC6B595Full{TPIC6B595Minimal: min}, nil
}

// Clear pulses SRCLR to clear the shift register only. The storage
// register (and therefore the DRAIN outputs) keeps its last-latched
// value until the next RCK pulse.
func (d *TPIC6B595Full) Clear() error {
	return d.sipo.Clear()
}

// SetOutputEnable drives G LOW (enabled=true) or HIGH (enabled=false).
// enabled=false forces every output off without disturbing the shadow
// register or the storage register's contents.
func (d *TPIC6B595Full) SetOutputEnable(enabled bool) error {
	return d.sipo.SetOutputEnable(enabled)
}

// WriteAll writes every cascaded device's byte in one call. values is
// length-truncated or zero-extended to numDevices as needed.
func (d *TPIC6B595Full) WriteAll(values []uint8) error {
	for i := uint8(0); i < d.numDevices; i++ {
		var v uint8
		if int(i) < len(values) {
			v = values[i]
		}
		d.shadow[i] = v
	}
	return d.flush()
}
